hummingbot/market/kraken/kraken_api_order_book_data_source.py

```diff
@@ -0,0 +1,299 @@
+                        raise
+        except asyncio.TimeoutError:
+            self.logger().warning("WebSocket ping timed out. Going to reconnect...")
+            return
+        except ConnectionClosed:
+            return
+        finally:
+            await ws.close()
+
+    async def listen_for_trades(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        while True:
+            try:
+                trading_pairs: List[str] = await self.get_trading_pairs()
+                async with websockets.connect(KRAKEN_WS_URI) as ws:
+                    ws: websockets.WebSocketClientProtocol = ws
+                    for trading_pair in trading_pairs:
+                        subscribe_request: Dict[str, Any] = {
+                            "event": "subscribe",
+                            "pair": [trading_pair],
+                            "subscription": {"name": "trade"}
+                        }
+                        await ws.send(json.dumps(subscribe_request))
+
+                    async for raw_msg in self._inner_messages(ws):
+                        msg = ujson.loads(raw_msg)
+                        if "pong" in msg:
+                            await ws.send({"event": "ping"})
+                        if type(msg) is list:
+                            if len(msg) == 4:
+                                for data in msg[1]:
+                                    trade_message: OrderBookMessage = KrakenOrderBook.trade_message_from_exchange(data, metadata={"symbol": msg[3]})
+                                    output.put_nowait(trade_message)
+                        # Server heartbeat sent if no subscription traffic within 1 second (approximately)
+                        elif msg["event"] == 'heartbeat':
+                            continue
+                        else:
+                            self.logger().debug(f"Unrecognized message received from Kraken websocket: {msg}")
+            except asyncio.CancelledError:
+                raise
+            except Exception:
+                self.logger().error("Unexpected error with WebSocket connection. Retrying after 30 seconds...",
+                                    exc_info=True)
+                await asyncio.sleep(30.0)
+
+    async def listen_for_order_book_diffs(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        while True:
+            try:
+                trading_pairs: List[str] = await self.get_trading_pairs()
+                async with websockets.connect(KRAKEN_WS_URI) as ws:
+                    ws: websockets.WebSocketClientProtocol = ws
+                    for trading_pair in trading_pairs:
+                        subscribe_request: Dict[str, Any] = {
+                            "event": "subscribe",
+                            "pair": [trading_pair],
+                            "subscription": {"name": "book"}
+                        }
+                        await ws.send(json.dumps(subscribe_request))
```

Comment:
According to Kraken's Web Socket Documentation, you can subscribe to a list of trading pairs. So there is really no need for a `for-loop`. Also there seems to be a `depth` parameter that you should have included.

As such the the following would be sufficient when subscribing to an orderbook
```
subscribe_request: Dict[str, Any] = {
    "event": "subscribe",
    "pair": trading_pairs,
    "subscription": {
        "name": "book"
        "depth": 1000
    }
}
await ws.send(json.dumps(subscribe_request))
```